### ISSUE TITLE
Questions to port model in Keras

### ISSUE THREAD

**Issue by shkarupa-alex** [locuslab/trellisnet#3]
I'm trying to port TrellisNet to Keras and got some questions.

1. Why do you use custom weight normalization https://github.com/locuslab/trellisnet/blob/master/TrellisNet/optimizations.py#L190 instead of common once
https://pytorch.org/docs/stable/_modules/torch/nn/utils/weight_norm.html ?
What is wrong with PyTorch core implementation?

2. Why do you manually recompute weights norm in upper layer  
https://github.com/locuslab/trellisnet/blob/master/TrellisNet/trellisnet.py#L145 instead of putting it inside `forward` method of WeightNorm itself?

3. Why do you use `dialation + device` key here https://github.com/locuslab/trellisnet/blob/master/TrellisNet/trellisnet.py#L55 instead of just `dialation`

4. It seems that `VariationalDropout` here https://github.com/locuslab/trellisnet/blob/master/TrellisNet/optimizations.py#L119 is not a true  `VariationalDropout` but simply https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/layers/SpatialDropout1D

**Comment by shaojieb** [comment]
Hi! Thanks for your interest in our work, and thank you for helping port it to Keras :-) Regarding your questions:

1. The PyTorch weight normalization recomputes the weight using `weight_g` and `weight_v` every time the module is called. In a TrellisNet where the convolutional filter is tied across all layers, this would mean we need to repeat this (re-)computation by L times for an L-layer TrellisNet. Therefore, I modified the code to make it more computationally efficient by just do the re-computation once (the reasoning is explained in the comment here too: https://github.com/locuslab/trellisnet/blob/master/TrellisNet/optimizations.py#L238). There is nothing wrong with the original PyTorch core implementation and you would get the same result.

2. This is a consequence of the point above. Since I would like to avoid "automatic" recomputations at every layer (that PyTorch does), I can simply recompute once at the start of the stacked layers in every training iterations. In other words, I only need to recompute when the weights are updated.

3. This is just to handle a PyTorch related problem when in a multi-gpu setting. When one uses `nn.DataParallel`, the dict aliasing could lead to a potential problem where GPU 0 accesses the tensor stored by GPU 1. There are other ways to bypass this issue (e.g., not using `dict`), so you might want to consider other options, too.

4. You are right that it is the same as the `SpatialDropout1D` when applied on a sequence. However, in RNNs, the variational dropout has two implications: 1) the same dropout mask is applied at every time step (of a sequence) when going from RNN layer i to i+1; and 2) the same dropout mask is applied when transitioning from time step t to t+1 (see https://arxiv.org/pdf/1512.05287.pdf). Given the relationship between TrellisNets and RNNs, the name `VariationalDropout` is just to remind the users that, once the dropout mask is created, this SpatialDropout1D should not only be applied on the sequence of a single layer, but on **all layers** of the deep network. In contrast, typically the SpatialDropout1D mask is different at each layer of a ConvNet.

Hope this helps! Please let me know if you have any other question, or if you need any help while porting TrellisNet to Keras.
